### DISCUSSION TITLE
How to use to use dynamic response_model when using SQLModel classes defined at runtime

### DISCUSSION THREAD

**Discussion by gsouveton** [fastapi/fastapi#6178] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List, Optional
from fastapi import Depends, FastAPI, HTTPException, Query
from sqlmodel import Field, Relationship, Session, SQLModel, create_engine, select
from sqlalchemy.orm import registry

all_bases = {}

class BaseRepository:
    def __init__(self, db_filename):
        self.engine = create_engine(
            f"sqlite:///{db_filename}",
            echo=True
        )
        self.mapper_registry = registry()
        class SQLModel_custom(SQLModel, registry=self.mapper_registry):
            pass
        self.sqlmodel_subclass = SQLModel_custom
        self.sqlmodel_subclass.metadata = self.mapper_registry.metadata


def generate_models(base_repository: BaseRepository):
    class TeamBase(base_repository.sqlmodel_subclass):
        name: str = Field(index=True)
        headquarters: str

    class Team(TeamBase, table=True):
        id: Optional[int] = Field(default=None, primary_key=True)
        heroes: List["Hero"] = Relationship(back_populates="team")

    class TeamCreate(TeamBase):
        pass

    class TeamRead(TeamBase):
        id: int

    class HeroBase(base_repository.sqlmodel_subclass):
        name: str = Field(index=True)
        secret_name: str
        age: Optional[int] = Field(default=None, index=True)
        team_id: Optional[int] = Field(default=None, foreign_key="team.id")

    class Hero(HeroBase, table=True):
        id: Optional[int] = Field(default=None, primary_key=True)
        team: Optional["Team"] = Relationship(back_populates="heroes")

    class HeroRead(HeroBase):
        id: int

    class HeroCreate(HeroBase):
        pass

    class HeroReadWithTeam(HeroRead):
        team: Optional["TeamRead"] = None

    class TeamReadWithHeroes(TeamRead):
        heroes: List["HeroRead"] = []
        
    base_repository.TeamBase = TeamBase
    base_repository.Team = Team
    base_repository.TeamCreate = TeamCreate
    base_repository.TeamRead = TeamRead
    base_repository.HeroBase = HeroBase
    base_repository.Hero = Hero
    base_repository.HeroRead = HeroRead
    base_repository.HeroCreate = HeroCreate
    base_repository.HeroReadWithTeam = HeroReadWithTeam
    base_repository.TeamReadWithHeroes = TeamReadWithHeroes

def get_base_repository(basename):
    return all_bases[basename]

def create_db_and_tables():
    for base in all_bases.values():
        generate_models(base)
        base.sqlmodel_subclass.metadata.create_all(base.engine)

def get_session(basename):
    with Session(all_bases[basename].engine) as session:
        yield session
        
app = FastAPI()

all_bases['AAA'] = BaseRepository('./aaa.db')
all_bases['BBB'] = BaseRepository('./bbb.db')
# and so on ... (base names and path are not static)
        

@app.on_event("startup")
def on_startup():
    create_db_and_tables()

@app.get("/{basename}/heroes/") # I wish I could put something like 'List[base_repository.HeroRead]' here
def read_heroes(
    *,
    base_repository = Depends(get_base_repository)
    session: Session = Depends(get_session),
):
    heroes = session.exec(select(base_repository.Hero)).all()
    return heroes
    
    
@app.get("/{basename}/heroes/{hero_id}") # I wish I could put something like 'base_repository.HeroRead' here
def read_hero(
    *,
    base_repository = Depends(get_base_repository),
    session: Session = Depends(get_session),
    hero_id: int
):
    hero = session.get(Hero, hero_id)
    if not hero:
        raise HTTPException(status_code=404, detail="Hero not found")
    return hero
```

### Description

Hello !

(Note : I'm using SQLModel but I think my issue is more generally related to FastAPI and the way it works with response models).

I'm working on an API that is linked to several databases and in which SQLModel classes are dynamically generated at runtime for each database with their own registry/metadata/schema.
It works well (although I had a hard time understanding how to overcome the "sqlalchemy.exc.InvalidRequestError: Multiple classes found for path "xxx" in the registry of this declarative base. Please use a fully module-qualified path.". See this issue : https://github.com/tiangolo/sqlmodel/issues/264).

My MRE is based on the code from this tutorial page : https://sqlmodel.tiangolo.com/tutorial/fastapi/relationships/
I added code so that the models are dynamically generated, just like in my real usecase.
The problem is that, from what I understand, the integration between FastAPI and SQLModel relies on 'response_model' parameter is used to automatically "cast" the data output from the select() into the desired returned type. (In my example that would be cast 'List[Hero]' into 'List[HeroRead]' or 'Hero' into 'HeroReadWithTeam'). 

However, in my case, data models are defined at runtime and don't exist yet when the routes are defined. It is not possible to use Union types either. But maybe is there an alternative way to define this response_model within the route body ?

Otherwise, if the response_model can not be dynamically defined, an alternative solution could be reimplementing such a type-casting feature, which seems quite difficult though and not the smartest way of solving this issue ; maybe is it possible to directly call FastAPI underlying mechanism ?

### Operating System

Linux

### FastAPI Version

0.86.0

### Python Version

3.9.5

**Comment by gsouveton** [comment, marked as answer]
Ok, I just realised I forgot about the existence of the magic "from_orm" method.
I just managed to solve my issue like so :

```
@app.get("/{basename}/heroes/")
def read_heroes(
    *,
    base_repository = Depends(get_base_repository)
    session: Session = Depends(get_session),
):
    heroes = session.exec(select(base_repository.Hero)).all()
    return [base_repository.HeroReadWithTeam.from_orm(h) for h in heroes]
    
    
@app.get("/{basename}/heroes/{hero_id}")
def read_hero(
    *,
    base_repository = Depends(get_base_repository),
    session: Session = Depends(get_session),
    hero_id: int
):
    hero = session.get(Hero, hero_id)
    if not hero:
        raise HTTPException(status_code=404, detail="Hero not found")
    return base_repository.HeroReadWithTeam.from_orm(hero)
```

**Comment by tiangolo** [comment]
Thanks for coming back to close the issue! ☕

**Comment by gsouveton** [comment]
Using 'from_orm' is indeed the right way of proceeding, but it introduces a new error :

`pydantic.errors.ConfigError: field "team" not yet prepared so type is still a ForwardRef, you might need to call HeroReadWithTeam.update_forward_refs().`

However, it is not possible to use 'update_forward_refs()' since it will look for a 'TeamRead' model that does not exist :

```
  File "pydantic/main.py", line 780, in pydantic.main.BaseModel.update_forward_refs
  File "pydantic/typing.py", line 480, in pydantic.typing.update_model_forward_refs
  File "pydantic/typing.py", line 450, in pydantic.typing.update_field_forward_refs
  File "pydantic/typing.py", line 63, in pydantic.typing.evaluate_forwardref
  File "/usr/lib/python3.9/typing.py", line 518, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'TeamRead' is not defined
```

which seems logical since there's no such 'TeamRead' globally defined.
The solution is to use 'update_forward_refs()' right after all classes have been defined in 'generate_models' function, and provide it a mapping to the locally defined classes like so :

```
    def generate_models(base_repository: BaseRepository):
    
        class TeamBase(base_repository.sqlmodel_subclass):
            name: str = Field(index=True)
            headquarters: str

        class Team(TeamBase, table=True):
            id: Optional[int] = Field(default=None, primary_key=True)
            heroes: List["Hero"] = Relationship(back_populates="team")

        class TeamCreate(TeamBase):
            pass

        class TeamRead(TeamBase):
            id: int

        class HeroBase(base_repository.sqlmodel_subclass):
            name: str = Field(index=True)
            secret_name: str
            age: Optional[int] = Field(default=None, index=True)
            team_id: Optional[int] = Field(default=None, foreign_key="team.id")

        class Hero(HeroBase, table=True):
            id: Optional[int] = Field(default=None, primary_key=True)
            team: Optional["Team"] = Relationship(back_populates="heroes")

        class HeroRead(HeroBase):
            id: int

        class HeroCreate(HeroBase):
            pass

        class HeroReadWithTeam(HeroRead):
            team: Optional["TeamRead"] = None

        class TeamReadWithHeroes(TeamRead):
            heroes: List["HeroRead"] = []

        base_repository.TeamBase = TeamBase
        base_repository.Team = Team
        base_repository.TeamCreate = TeamCreate
        base_repository.TeamRead = TeamRead
        base_repository.HeroBase = HeroBase
        base_repository.Hero = Hero
        base_repository.HeroRead = HeroRead
        base_repository.HeroCreate = HeroCreate
        base_repository.HeroReadWithTeam = HeroReadWithTeam
        base_repository.TeamReadWithHeroes = TeamReadWithHeroes
        
        entities = {}
        entities['TeamBase'] = TeamBase
        entities['Team'] = Team
        # ...
         
        base_repository.TeamBase.update_forward_refs(**entities)
        base_repository.Team.update_forward_refs(**entities)
        # ...
```

**Comment by tiangolo** [comment]
And does calling `update_forward_refs()` work?

**Comment by gsouveton** [comment]
`update_forward_refs()` with no parameter does not work since it will look for globally defined names and classes defined in the repository local namespace aren't.
But by providing the mapping as parameter (either explicitly : `update_forward_refs(TeamBase=base_repository.TeamBase, Team=base_repository.Team, ...)` or with a dict mapping like I did in my example above) it works fine!

**Comment by tiangolo** [comment]
Awesome! Thanks for reporting back. 🤓
